### PR TITLE
[SPARK-4172] [PySpark] Progress API in Python

### PR DESCRIPTION
**Pull request by davies**
This patch bring the pull based progress API into Python, also a example in Python.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Looks like this failed due to Scala style issues due to a few long lines.

**Comment by davies** [comment]
@JoshRosen I fixed it locally, but forgot to push out.

File: core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala [hunk 1]
```diff
@@ -95,25 +96,30 @@ private[spark] class TaskResultGetter(sparkEnv: SparkEnv, scheduler: TaskSchedul
   def enqueueFailedTask(taskSetManager: TaskSetManager, tid: Long, taskState: TaskState,
     serializedData: ByteBuffer) {
     var reason : TaskEndReason = UnknownReason
-    getTaskResultExecutor.execute(new Runnable {
-      override def run(): Unit = Utils.logUncaughtExceptions {
-        try {
-          if (serializedData != null && serializedData.limit() > 0) {
-            reason = serializer.get().deserialize[TaskEndReason](
-              serializedData, Utils.getSparkClassLoader)
+    try {
```
Comment (davies) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala]: This is not related change, it mute the exception when you cancel a job.

**Comment by JoshRosen** [comment]
LGTM, so I'm going to merge this into `master` (1.4.0) and `branch-1.3` (1.3.0).  Thanks!
